### PR TITLE
Initial implementation of the Primare Binding

### PR DESCRIPTION
**Pull request by vpjuslin**
Documentation: https://github.com/vpjuslin/openhab/wiki/Primare-binding

### CODE REVIEW

**Comment by teichsta** [comment]
Thanks @vpjuslin for this contribution. I hope to find the time for reviewing it soon. Thanks, Thomas E.-E.

File: bundles/binding/org.openhab.binding.primare/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ org.openhab.core.library.items,
+ org.openhab.core.library.types,
+ org.openhab.core.types,
+ org.openhab.model.item.binding,
+ org.osgi.framework,
+ org.osgi.service.cm,
+ org.osgi.service.component,
+ org.osgi.service.event,
+ org.slf4j
+Export-Package: org.openhab.binding.primare
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.primare/META-INF/MANIFEST.MF]: remove this export unless any other bundle needs access?

File: bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/PrimareBinding.java [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+	private static final Pattern EXTRACT_CONFIG_PATTERN = Pattern.compile("^(.*?)\\.(host|port|serial|model)$");
+	
+	/** Map table to store all available receivers configured by the user */
+	protected Map<String, DeviceConfig> deviceConfigCache = null;
+	
+	public PrimareBinding() {
+	}
+
+	public void activate() {
+		logger.debug("Activate");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/PrimareBinding.java]: remove

File: bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/PrimareBinding.java [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+	}
+
+	public void deactivate() {
+		logger.debug("Deactivate");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/PrimareBinding.java]: remove

File: bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/protocol/PrimareSerialConnector.java [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+			} else {
+				logger.error("There are no communications ports available");
+			}
+			logger.error("You may consider OpenHAB startup parameter [ -Dgnu.io.rxtx.SerialPorts={} ]", 
+				     serialPortName);
+			throw e;
+			
+			//	} catch (Exception e) {
+			//logger.error("Error connecting to {}: {}", PrimareSerialConnector.this.toString(),e);
+			//throw e;
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/protocol/PrimareSerialConnector.java]: remove commented code

File: bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/protocol/PrimareResponse.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	
+	/**
+	 * Check if a response message from Primare is relevant for a Primare command.
+         * Since responses arrive asynchronously
+	 * 
+         * @param deviceCmdString
+         *              Primare command as specified in items configuration file
+	 * @return boolean
+	 */
+	public abstract boolean is_relevant_for(String deviceCmdString);
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/protocol/PrimareResponse.java]: according to the Java Coding Conventions rename this method to isRelevantFor

File: bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/protocol/PrimareSerialConnector.java [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+				     PrimareSerialConnector.this.toString());
+			
+			while (true) {
+				logger.trace("waitStateMessages - waiting data from {}",
+					     PrimareSerialConnector.this.toString());
+				byte b = inStream.readByte();
+				bytesReceivedAt = new Date();
+				
+				// logger.trace("waitStateMessages - data from {} READ: [{}] ({})",
+				//	     PrimareSerialConnector.this.toString(), String.format("0x%02x",b), b);
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/protocol/PrimareSerialConnector.java]: remove commented code

**Comment by teichsta** [comment]
Hi @vpjuslin,

please find some minor review comments inline. Please finally squash the commits into one once you've incorporated my feedback.

Thanks, Thomas E.-E.

File: bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/PrimareBinding.java (hunk 3 above)
Comment (vpjuslin) [inline bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/PrimareBinding.java]: Could you please elaborate?

Should the following be removed:

1. The constructor which is effectively the default constructor:

public PrimareBinding() {
}


2. activate()

public void activate() {
   logger.debug("Activate");
}


3. deactivate()

public void deactivate() {
   logger.debug("Deactivate");
   closeAllConnections();
}


Or do you mean just the logger.debug() statements?


/vp

Comment (teichsta) [inline bundles/binding/org.openhab.binding.primare/src/main/java/org/openhab/binding/primare/internal/PrimareBinding.java]: just meant the log statement

**Comment by vpjuslin** [comment]
Sorry to bother you again (this is somewhat new to me)

Q1: Do you prefer all my Primare binding -related commits to be squashed 
or just the commits related to the review?

Q2: Is it ok to do the squashing with:

git reset --soft <parent-of-my-first-commit>
git commit

followed by a

git push --force

or does using the --force option interfere with the review process? Is 
there a preferred option?


/vp

**Comment by teichsta** [comment]
> Q1: Do you prefer all my Primare binding -related commits to be squashed or just the commits related to the review?

All commits please. The whole PR should come with just one single commit.

> Q2: Is it ok to do the squashing with:

never used that workflow. For now must of the contributors followed the workflow described [here](https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit)

**Comment by vpjuslin** [comment]
For some reason my commit history was so complicated that I did not know 
how to squash it into a single commit, so I created a new pull request 
instead. Sorry about that.

/vp

**Comment by teichsta** [comment]
superseded by #2625

**Pull request closed without merging**